Windows: Cpack wix installer

adrido opened the pull request:
This allows CMake/CPack to use the Windows Installer XML (WIX) to create a Installer package.

The Installer package is only available if its **not** a `RUN_IN_PLACE` build. For that the zip package is generated.

Todo:
- [ ] ~~Uninstaller should clean the cache dir.~~ *
- [ ] ~~Question: should uninstaller also remove the minetest.conf?~~ *
- [x] Each sub-game should have its own component
    - [x] "minimal" sub-game should may not installed by default only optional

*= Applications are installed for all users, while config file and cache is per user. Removing this while uninstalling is neither a good idea nor supported by MSI.
Related: #4931, #3596
Fixes: #4528, #3802, #4697
### Download:
[Minetest-0.5.0-win32.msi.zip](https://github.com/minetest/minetest/files/1777352/Minetest-0.5.0-win32.msi.zip)
[Minetest-0.5.0-win64.msi.zip](https://github.com/minetest/minetest/files/1777354/Minetest-0.5.0-win64.msi.zip)

The installer here is packed in a zip, because Github does not allow MSI files.

### Screenshots:
![grafik](https://user-images.githubusercontent.com/3142688/28356648-206149ee-6c69-11e7-940f-52528cd84e9c.png)
![grafik](https://user-images.githubusercontent.com/3142688/28356661-30aa9166-6c69-11e7-99ee-12eabdba36c3.png)
![grafik](https://user-images.githubusercontent.com/3142688/28708876-2395f2d8-737e-11e7-9c69-0eaef63d1c13.png)
The minimal subgame have to be manually selected.
![grafik](https://user-images.githubusercontent.com/3142688/28356710-509dceac-6c69-11e7-9a65-af004f9cd50e.png)
![grafik](https://user-images.githubusercontent.com/3142688/28356726-61a96800-6c69-11e7-9676-9046e438d07c.png)
![grafik](https://user-images.githubusercontent.com/3142688/28356734-66a0f6fc-6c69-11e7-9c64-5dc8396ffb8d.png)

nerzhul commented:
interesting !

adrido commented:
Should be ready for review now.

SmallJoker commented on CMakeLists.txt:
```diff
@@ -217,14 +218,61 @@ set(CPACK_PACKAGE_VERSION_PATCH ${VERSION_PATCH})
+		set(CPACK_CREATE_DESKTOP_LINKS ${PROJECT_NAME})
+
+		set(CPACK_WIX_PRODUCT_ICON "${CMAKE_CURRENT_SOURCE_DIR}/misc/minetest-icon.ico")
+		# Supported languages can be found at http://wixtoolset.org/documentation/manual/v3/wixui/wixui_localization.html
```
Line too long

SmallJoker commented on CMakeLists.txt:
```diff
@@ -217,14 +218,61 @@ set(CPACK_PACKAGE_VERSION_PATCH ${VERSION_PATCH})
+		
+		set(CPACK_RESOURCE_FILE_LICENSE "${CMAKE_CURRENT_SOURCE_DIR}/doc/lgpl-2.1.txt")
+
+		# The correct way would be to include both x32 and x64 into one installer and install the approitive one.
```
Too long here too

SmallJoker commented on README.txt:
```diff
@@ -399,6 +399,21 @@ popd
+
+In Visual Studio 2017 Installer select "optional Features" -> "Wix Toolset"
+
+Build the binaries like described above, but make shure you unselect "RUN_IN_PLACE".
```
* make sure

SmallJoker commented on README.txt:
```diff
@@ -399,6 +399,21 @@ popd
+In Visual Studio 2017 Installer select "optional Features" -> "Wix Toolset"
+
+Build the binaries like described above, but make shure you unselect "RUN_IN_PLACE".
+"RUN_IN_PLACE" builds are not designed for installations and leads to errors if installd on newer Windows since XP!
```
\* make sure
`and leads to errors.` (shorten and remove "installed" misspelling)

sfan5 commented on CMakeLists.txt:
```diff
@@ -217,14 +218,61 @@ set(CPACK_PACKAGE_VERSION_PATCH ${VERSION_PATCH})
+		# Supported languages can be found at http://wixtoolset.org/documentation/manual/v3/wixui/wixui_localization.html
+		#set(CPACK_WIX_CULTURES "ar-SA,bg-BG,ca-ES,hr-HR,cs-CZ,da-DK,nl-NL,en-US,et-EE,fi-FI,fr-FR,de-DE")
+		set(CPACK_WIX_UI_BANNER "${CMAKE_CURRENT_SOURCE_DIR}/misc/CPACK_WIX_UI_BANNER.BMP")
+		set(CPACK_WIX_UI_DIALOG "${CMAKE_CURRENT_SOURCE_DIR}/misc/CPACK_WIX_UI_DIALOG.BMP")
```
the files are not included in the pr

adrido replied:
You may have overseen them, they are included indeed: 
https://github.com/adrido/minetest/blob/03bbfb935fc86da367d48bf6ae864c9a5da1aea9/misc/CPACK_WIX_UI_BANNER.BMP
https://github.com/adrido/minetest/blob/03bbfb935fc86da367d48bf6ae864c9a5da1aea9/misc/CPACK_WIX_UI_DIALOG.BMP

Yes, they need to be BMP files. And they need to have exactly that resolution.

sfan5 replied:
oh, my mistake

sfan5 commented on CMakeLists.txt:
```diff
@@ -217,14 +218,61 @@ set(CPACK_PACKAGE_VERSION_PATCH ${VERSION_PATCH})
 set(CPACK_PACKAGE_VENDOR "celeron55")
 set(CPACK_PACKAGE_CONTACT "Perttu Ahola <celeron55@gmail.com>")
 
+include(CPackComponent)
```
the Docs component should have a display name of `Documentation`

adrido replied:
Yes of course. Thanks. Its done.

sfan5 commented on src/CMakeLists.txt:
```diff
@@ -727,8 +727,8 @@ if(MSVC)
 	# Flags for C files (sqlite)
-	# /MT = Link statically with standard library stuff
-	set(CMAKE_C_FLAGS_RELEASE "/O2 /Ob2 /MT")
+	# /MD = Link statically to MSVCRT.lib and dynamically to MSVCR<version>.DLL
```
this is worded confusingly, just say `dynamically link to MSVCR dll`

sfan5 commented on src/porting.cpp:
```diff
@@ -353,14 +353,21 @@ bool setSystemPaths()
 	FATAL_ERROR_IF(len == 0 || len > sizeof(buf), "Failed to get APPDATA");
 
-	path_user = std::string(buf) + DIR_DELIM + PROJECT_NAME;
+	path_user = std::string(buf) + DIR_DELIM + PROJECT_NAME_C;
```
any good reason for this change?

adrido replied:
Find the wrong one: 😉 
![grafik](https://user-images.githubusercontent.com/3142688/28752102-e4561e90-7517-11e7-95a6-defaea022dc3.png)

sfan5 commented on src/porting.cpp:
```diff
@@ -353,14 +353,21 @@ bool setSystemPaths()
-	path_share = std::string(buf) + "\\..";
+	path_share = exepath + "\\..";
+	if (detectMSVCBuildDir(exepath)) {
+		// The msvc build dir schould normaly not present if properly installed,
```
... `not be present` ...

sfan5 commented on CMakeLists.txt:
```diff
@@ -217,14 +218,61 @@ set(CPACK_PACKAGE_VERSION_PATCH ${VERSION_PATCH})
+	# Include all dynamically linked runtime libaries
+	include(InstallRequiredSystemLibraries)
+
+	if(RUN_IN_PLACE)
```
since this whole process seems to require building on Windows,
shouldn't it still generate ZIPs for RUN_IN_PLACE=0 compiled via MinGW?

adrido replied:
This would not make much sense. 

Also this would cause much confusion to players if accidentally shared.
e.g. forum posts like this would occur:
"PLASE HELP Mods don't load", "Where is my minetest.conf?"

It does not explicit requires building with MSVC, maybe even MinGW is supported. There is a standalone package available: http://wixtoolset.org/ 
It just depends on CMake to support that. (not tested)

sfan5 replied:
So what happens if I compile using MinGW <b>on Linux</b> (-> no WiX toolset) with RUN_IN_PLACE=0 and run `make package`?

> Also this would cause much confusion to players if accidentally shared.

Why? It would use the exact same runtime paths as if it was installed using the installer.

adrido replied:
> So what happens if I compile using MinGW on Linux (-> no WiX toolset) with RUN_IN_PLACE=0 and run make package?

a) nothing
b) it prints an error message
c) it prints a warning and generates a zip package anyway
d) it generates a zip package without any other message
e) it downloads and installs the wix toolset and generates a msi package

Well I don't know. 😆  How did you generate your 7zip packages you uploaded [here](https://minetest.kitsunemimi.pw/builds/)? You could probably do the same only as zip format.

> Why? It would use the exact same runtime paths as if it was installed using the installer.

The problem is to tell the kiddies that "zip package 1" finds its mods and conf in a different place then "zip package 2" 🤔

sfan5 replied:
> Well I don't know. :laughing: How did you generate your 7zip packages you uploaded here? You could probably do the same only as zip format.

I use `make package` but that's not the point.
The problem I have here is that you are willfully breaking package creation when cross-compiling for Windows with RUN_IN_PLACE=0

(also since you didn't seem to understand: the WiX toolset does <b>not</b> work on Linux, like not at all)


> The problem is to tell the kiddies that "zip package 1" finds its mods and conf in a different place then "zip package 2" :thinking:

That's true I guess.
But IMO trying to protect devs from building broken packages is not useful, previously the build system did not prevent building RUN_IN_PLACE=0 ZIP packages either.

Calinou commented:
Note that it's not required for the users to accept the LGPL license in order to play Minetest, so the "license" step of the installer can be removed (to make installation a bit quicker).

adrido commented:
> Note that it's not required for the users to accept the LGPL license in order to play Minetest, so the "license" step of the installer can be removed

The problem is, if no licence file is speciefied, CMake generates and shows a dummy file. This reads like "The package creator was too lazy to choose a real license". So its imo the best to show users that Minetest is open source, licensed under LGPL. Other OS programs shows the GPL or LGPL license file too while installing e.g. Gimp

sfan5 commented on CMakeLists.txt:
```diff
@@ -217,14 +218,68 @@ set(CPACK_PACKAGE_VERSION_PATCH ${VERSION_PATCH})
-	if(CMAKE_SIZEOF_VOID_P EQUAL 8)
-		set(CPACK_PACKAGE_FILE_NAME "${PROJECT_NAME}-${VERSION_STRING}-win64")
+	# Include all dynamically linked runtime libaries
+	include(InstallRequiredSystemLibraries)
```
DLL files are already installed manually, what's this for?

adrido replied:
that installs the correct msvcr dlls automatically. No need to add them manually anymore :-)

sfan5 commented on README.txt:
```diff
@@ -399,6 +404,23 @@ popd
+In Visual Studio 2017 Installer select "optional Features" -> "Wix Toolset"
+
+Build the binaries like described above, but make sure you unselect "RUN_IN_PLACE".
+"RUN_IN_PLACE" builds are not designed for installations and leads to errors
```
this is also confusing because creating RUN_IN_PLACE=1 installer builds is impossible (prevented by the build system)

adrido replied:
I can remove the last sentence if think its better.

adrido commented:
Rebased.

adrido commented:
Rebased.

adrido commented:
> Looks good. I'm not on windows, so I can't test this. You need to make sure that any writing files (games, worlds, mods, config) can be written correctly.

The games folder (minetest_game) is protected, and cannot be written. The config file, worlds folder and mods folder can be written. User installed games inside the shared folder can be written too. Is this behavior wrong?

I guess the main problem with this PR is, that all people that are allowed to approve are using Linux 😞 and cannot test this. I hope someone is interested enough to install Windows in a VM to be able to test and approve this PR.

SmallJoker commented on README.md:
```diff
@@ -88,6 +88,10 @@ Where each location is on each platform:
+* Windows installed:
+    * $bin   = C:\Program Files\Minetest\bin
+    * $share = C:\Program Files\Minetest
+    * $user  = C:\Users\<user>\AppData\Roaming\Minetest
```
Shouldn't we rather use something like `%APPDATA%\Minetest` respectively `%PROGRAMFILES%\Minetest` for these paths?

adrido replied:
As you can see in https://github.com/minetest/minetest/pull/6153/files#diff-59f99d02fe5b328bb9a57b023497b5edR367 It does automatically resolve the %APPDATA% path. 
And the Installer does also resolve the %PROGRAMFILES% variable and uses this as default.

This is only documentation, but i think a 10 year old will better understand the full written path instead of %APPDATA%\Minetest. A experienced user that changed the Environment variable will also understand that his data is then in his custom path. But I can also change this part of documentation, its on your opinion.

Calinou replied:
`%PROGRAMFILES%\Minetest` and `%APPDATA%\Minetest` can be entered in the Windows Explorer path field (or in the `Windows+R` Run dialog) and Windows will replace the variable and access the correct path.

adrido replied:
Im sorry, I changed this some days ago, but It did not sucessfully push (probably due to network trouble).

I only changed $user to %Appdata%\Minetest because %ProgramFiles%\Minetest will not always resolve to the correct path:
 a) User installed it somewhere else
 b) User installed the x86 version on x64 Windows (WOW64)

adrido commented:
Rebased, updated x64 installer, added x86 installer.

Fixer-007 commented:
Half a year passed, it is 2018, roses are not yet red, nobody interested yet?

adrido commented:
Thanks for being the first person, that tries the x32 build after a 3/4 Year. I apologize that it was broken. That was related to the build system, not to this PR. 
I rebased this pr and updated the packages. I hope it gets merged before its 1st anniversary.
Now to your points:
1) Installing programs always requires advanced privileges, even if you have full access to the install destination, because the installer creates an entry in the "Installed Program's" list. This is done using registry entries. Modifying the registry always requires administrative privileges.  Also Installations are always made system wide, for all users, per user Installations were removed since Windows Vista or 7.
2) Was caused by a broken build of zlib. curl could not uncompress the serverlist response. I should have noticed this much earlier, but I thought this is just another failure of the master-server list.
3) That's expected. :-)
4) That was also zlib's vault.
5) My builds does not have gettext enabled, because of licensing trouble: Gettext does not allow to be shared together with Microsoft redistributables. Another reason is, that its nearly impossible to build with MSVC
6) I would hope so. The  %appdata% folder is guaranteed to be always writable
7) As I wrote in the first post, This is not a good idea nor supported by wix. Installations are made system wide, while the %appdata% folder is per user. How should the Installer know which users want to remover their files and which not? Maybe there can be a manual solution implemented, but if even this PR gets no approvals how should it get approvals if it have more features that have to be tested?

Anyway thanks for your interest and your tests.

Fixer-007 commented:
BTW, why it says this feature requires 0 kb on screenshot N3?

Calinou commented:
> My builds does not have gettext enabled, because of licensing trouble: Gettext does not allow to be shared together with Microsoft redistributables.

Can you please link justification for that? Microsoft redistributables fall under the "system library exception" as defined by the (L)GPL.

adrido commented:
> BTW, why it says "this feature requires 0 kb" on screenshot N3?

I had already installed it, and reinstalled it with a installer that had a different signature/timestamp/whatever but 0 bytes difference. 8)

> This is not true; you can create user-wide Start Menu entries in %APPDATA%\Microsoft\Windows\Start Menu\Programs. See my Godot Inno Setup installer for an example of an installer that doesn't ask for administrator privileges.

I guess you did misunderstand me. The "Installed Programs" is not the list in the startmenu. Its the list displayed in system control under "Add and remove Programs" or "All Apps" on Windows 10.

> Can you please link justification for that? As far as I know, Microsoft C++ redistributables fall under the "system library" exception as defined by the (L)GPL.

https://www.gnu.org/licenses/gpl-faq.en.html#WindowsRuntimeAndGPL
2nd paragraph.
Im not a lawyer, but as far I understood it, the redistributables in their dll form are not system libraries.
But discussing this here would be offtopic.

Fixer-007 commented:
I've rechecked 32 bit installer, works fine to me, installing textures/games/mods into %appdata%/minetest works, few wishes:
+ create empty folders in %appdata%/minetest for assets, aka games, mods, textures, so player can recognise were to put addons
+ locales support
+ I'm still convinced you can install without admin privs, windows 7 even can add it to "Programs" list

Calinou commented:
> I'm still convinced you can install without admin privs, windows 7 even can add it to "Programs" list

Indeed, my Godot installer appears in Programs and Features once installed.

Fixer-007 commented:
I guess MSI can't do this, but other installers can.

adrido commented:
Oh, I have to apologize, you both were right. MSI supports perUser installation and the WixToolset too. (http://wixtoolset.org/documentation/manual/v3/xsd/wix/package.html) The problem is, that CMake/CPack does not offer a switch for it (https://cmake.org/cmake/help/v3.10/module/CPackWIX.html). It would possible to workaround that using a patch file, but that would make it more complex and thus more unlikely to merge. 
Apart from that in the last years of working with Windows I had only one Installer that did a perUser installation: The whatsapp "native" desktop client. (I immediately uninstalled it.) 
The correct place for programs is still C:\Program Files\ and if someone want to use it only for himself, he can still use the portable version. 
If someone does not have administrative privileges, he may not allowed to install software neither a per user installation nor a per machine installation. If a child want to install it on the family computer it have to ask it parents for permissions (which is absolutely correct).

the multilanguage installer is a feature request since many years for cmake: https://gitlab.kitware.com/cmake/cmake/issues/14925

The empty folders may added later, first this have to be merged, everything else would wasted time. That may also a good idea for the linux users, since their installed version is also missing them.

rubenwardy commented:
I'd like to see this for 0.5.0. What remains to be done?

adrido commented:
From my side its ready. I would also like to see this in 0.4.17+

paramat commented:
Waiting for reviews.

adrido commented on CMakeLists.txt:
```diff
@@ -237,14 +239,68 @@ set(CPACK_PACKAGE_VERSION_PATCH ${VERSION_PATCH})
+cpack_add_component(SUBGAME_MINIMAL
+	DISPLAY_NAME "Minimal development test"
+	DESCRIPTION "A minimal subgame helping to develop the engine."
+	DISABLED #DISABLED does not mean it is disabled, it means its not selected on default!
```
In my opinion that comment is really important, because if something is disabled (e.g. a button, or in this case a checkbox) it means it is grayed out and not accessible/changeable/clickable by the user. But for cmake it means just unchecked
